metals/src/main/scala/scala/meta/internal/metals/callHierarchy/CallHierarchyProvider.scala

```diff
@@ -0,0 +1,619 @@
+                definitionInfo
+              definitionRange <- definitionOccurence.range
+              if !(definitionDoc == doc && realRoot.exists {
+                case (definition, name) =>
+                  !(name.pos.toLSP == definitionRange.toLSP) && definition.pos
+                    .encloses(definitionRange.toLSP)
+              })
+              definitionName <- trees.findLastEnclosingAt(
+                definitionSource,
+                definitionRange.toLSP.getStart,
+              )
+              (definition, _) <- getSpecifiedOrFindDefinition(
+                Some(definitionName)
+              )
+            } yield FindOutgoingCallsResult(
+              definitionOccurence,
+              definition.pos.toLSP,
+              List(name.pos.toLSP),
+              definitionSource,
+              definitionDoc,
+            )
+
+          } yield result).map(_.toList)
+        case t
+            if extractNameFromMember(t).isDefined || t.is[Term.Param] || t
+              .is[Pat.Var] =>
+          Future.successful(Nil)
+        case other =>
+          Future.sequence(other.children.map(search)).map(_.flatten)
+      }
+
+    realRoot match {
+      case Some((definition, _)) =>
+        Future
+          .sequence(
+            (definition match {
+              case member: Member =>
+                val nameDefinitionIndex = member.children.indexOf(member.name)
+                (if (nameDefinitionIndex == -1) member.children
+                 else member.children.patch(nameDefinitionIndex, Nil, 1))
+                  .map(search)
+              case other => List(search(other))
+            })
+          )
+          .map(results => FindOutgoingCallsResult.group(results.flatten))
+      case None => Future.successful(Nil)
+    }
+  }
+
+  private def findOutgoingCallsSynthetics(
+      source: AbsolutePath,
+      doc: TextDocument,
+      root: Tree,
+  )(implicit ec: ExecutionContext): Future[List[
+    FindOutgoingCallsResult
+  ]] =
+    (findRealRoot(root) match {
+      case Some((definition, _)) =>
+        val definitionRange = definition.pos.toSemanticdb
+        Future
+          .sequence(
+            doc.synthetics
+              .flatMap(syn =>
+                extractSelectTree(syn.tree).collect {
+                  case s.SelectTree(
+                        s.OriginalTree(Some(range)),
+                        Some(s.IdTree(symbol)),
+                      )
+                      if definitionRange
+                        .encloses(range) && getSpecifiedOrFindDefinition(
+                        trees.findLastEnclosingAt(source, range.toLSP.getStart),
+                        Some(definition),
+                      ).exists(_._1 == definition) =>
+                    lazy val mightBeCaseClassConstructor =
+                      """\.apply\(\)\.$""".r
+
+                    for {
+                      definitionInfo <- findDefinitionOccurence(source, symbol)
+                        .flatMap {
```

Comment:
Same thing as pretty much all the other comments, let's extract some of this stuff out. We have a _ton_ going on in here which makes it hard to read and follow.